dns-server/src/dns_server.rs

```diff
@@ -227,55 +270,70 @@ async fn handle_dns_message(
     let name = query.original().name().clone();
     let records = store.query(mr)?;
     let rb = MessageResponseBuilder::from_message_request(mr);
+    let mut additional_records = vec![];
     let response_records = records
         .into_iter()
-        .map(|record| match record {
-            DnsRecord::A(addr) => {
-                let mut a = Record::new();
-                a.set_name(name.clone())
-                    .set_rr_type(RecordType::A)
-                    .set_data(Some(RData::A(addr)));
-                Ok(a)
-            }
-
-            DnsRecord::AAAA(addr) => {
-                let mut aaaa = Record::new();
-                aaaa.set_name(name.clone())
-                    .set_rr_type(RecordType::AAAA)
-                    .set_data(Some(RData::AAAA(addr)));
-                Ok(aaaa)
-            }
-
-            DnsRecord::SRV(crate::dns_types::SRV {
-                prio,
-                weight,
-                port,
-                target,
-            }) => {
-                let tgt = Name::from_str(&target).map_err(|error| {
-                    RequestError::ServFail(anyhow!(
-                        "serialization failed due to bad SRV target {:?}: {:#}",
-                        &target,
-                        error
-                    ))
-                })?;
-                let mut srv = Record::new();
-                srv.set_name(name.clone())
-                    .set_rr_type(RecordType::SRV)
-                    .set_data(Some(RData::SRV(SRV::new(
-                        prio, weight, port, tgt,
-                    ))));
-                Ok(srv)
+        .map(|record| {
+            let record = dns_record_to_record(&name, record)?;
+
+            // DNS allows for the server to return additional records
+            // that weren't explicitly asked for by the client but that
+            // the server expects the client will want. The records
+            // corresponding to a lookup on a SRV target is one such case.
+            // We opportunistically attempt to resolve the target here
+            // and if successful return those additional records in the
+            // response.
+            // NOTE: we only do this one-layer deep.
+            if let Some(RData::SRV(srv)) = record.data() {
```

Comment:
I see you already pulled this into https://github.com/oxidecomputer/omicron/pull/3785/ , but out of curiosity, does it change the behavior of the rest of this PR? Are these additional records actually being read anywhere?

Reply:
No, as-is there are some bugs in `trust-dns` that block that, hence why I just pulled it out of this PR. I need to finish rebasing this one.